src/crypto/SessionKeystore.h

```diff
@@ -0,0 +1,103 @@
+#include <lib/support/Span.h>
+
+namespace chip {
+namespace Crypto {
+
+/**
+ * @brief Interface for deriving and managing lifecycle of session keys.
+ *
+ * This interface provides an abstraction that allows to store session keys in a secure
+ * environment. It uses the concept of key handles that isolate the application from the actual
+ * key material.
+ */
+class SessionKeystore
+{
+public:
+    virtual ~SessionKeystore() {}
+
+    /**
+     * @brief Import raw key material and return key handle.
+     *
+     * @note This method hould only be used when using the raw key material in the Matter stack
+     * cannot be avoided. Ideally, crypto interfaces should allow platforms to perform all the
+     * cryptographic operations in a secure environment.
+     *
+     * If the function returns no error, the application is responsible for destroying the handle
+     * using DestroyKey() method when the key is no longer needed.
+     */
+    virtual CHIP_ERROR CreateKey(const Aes128KeyByteArray & keyMaterial, Aes128KeyHandle & key) = 0;
+
+    /**
+     * @brief Derive key from shared secret.
+     *
+     * Use HKDF as defined in the Matter specification to derive an AES key from the shared secret.
+     *
+     * TODO: Replace P256ECDHDerivedSecret with an opaque HKDF master key.
+     *
+     * If the function returns no error, the application is responsible for destroying the handle
+     * using DestroyKey() method when the key is no longer needed.
+     */
+    virtual CHIP_ERROR DeriveKey(const P256ECDHDerivedSecret & secret, const ByteSpan & salt, const ByteSpan & info,
+                                 Aes128KeyHandle & key) = 0;
+
+    /**
+     * @brief Derive session keys from shared secret.
+     *
+     * Use HKDF as defined in the Matter specification to derive AES keys for both directions, and
+     * the attestation challenge from the shared secret.
+     *
+     * TODO: Replace ByteSpan for secret with an opaque HKDF master key.
+     *
+     * If the function returns no error, the application is responsible for destroying the handles
+     * using DestroyKey() method when the keys are no longer needed. On failure when deriving any of
+     * the keys, the method is expected to release all handles allocated so far.
+     */
+    virtual CHIP_ERROR DeriveSessionKeys(const ByteSpan & secret, const ByteSpan & salt, const ByteSpan & info,
+                                         Aes128KeyHandle & i2rKey, Aes128KeyHandle & r2iKey,
+                                         Aes128KeyByteArray & attestationChallenge) = 0;
+
+    /**
+     * @brief Destroy key.
+     *
+     * The method is considered a no-op when the argument is an uninitialized handle.
+     * After returning from the method the handle is expected to be deinitialized.
+     */
+    virtual void DestroyKey(Aes128KeyHandle & key) = 0;
+};
+
+/**
+ * @brief RAII class to release temporary key handle on scope exit.
+ */
+class SessionKeyGuard
```

Comment:
Suggest `AutoReleaseSessionKey` or `SafeTemporarySessionKey` as a name maybe?

Reply:
We seem to be overloading the term "safe" a bit ;) but `AutoReleaseSessionKey` sounds good to me.